modules/axdma/src/dma.rs

```diff
@@ -0,0 +1,129 @@
+        Self {
+            alloc: DefaultByteAllocator::new(),
+        }
+    }
+
+    /// Allocate arbitrary number of bytes. Returns the left bound of the
+    /// allocated region.
+    ///
+    /// It firstly tries to allocate from the coherent byte allocator. If there is no
+    /// memory, it asks the global page allocator for more memory and adds it to the
+    /// byte allocator.
+    pub unsafe fn alloc_coherent(&mut self, layout: Layout) -> AllocResult<DMAInfo> {
+        if layout.size() >= PAGE_SIZE_4K {
+            self.alloc_coherent_pages(layout)
+        } else {
+            self.alloc_coherent_bytes(layout)
+        }
+    }
+
+    fn alloc_coherent_bytes(&mut self, layout: Layout) -> AllocResult<DMAInfo> {
+        let mut is_expanded = false;
+        loop {
+            if let Ok(data) = self.alloc.alloc(layout) {
+                let cpu_addr = VirtAddr::from(data.as_ptr() as usize);
+                return Ok(DMAInfo {
+                    cpu_addr: data,
+                    bus_addr: virt_to_bus(cpu_addr),
+                });
+            } else {
+                if is_expanded {
+                    return Err(AllocError::NoMemory);
+                }
+                is_expanded = true;
+                let available_pages = global_allocator().available_pages();
+                // 4 pages or available pages.
+                let num_pages = 4.min(available_pages);
+                let expand_size = num_pages * PAGE_SIZE_4K;
+                let vaddr_raw = global_allocator().alloc_pages(num_pages, PAGE_SIZE_4K)?;
+                let vaddr = VirtAddr::from(vaddr_raw);
+                self.update_flags(
+                    vaddr,
+                    num_pages,
+                    MappingFlags::READ | MappingFlags::WRITE | MappingFlags::UNCACHED,
+                )?;
+                self.alloc
+                    .add_memory(vaddr_raw, expand_size)
+                    .inspect_err(|e| error!("add memory fail: {e:?}"))?;
+                debug!("expand memory @{vaddr:#X}, size: {expand_size:#X} bytes");
+            }
+        }
+    }
+
+    fn alloc_coherent_pages(&mut self, layout: Layout) -> AllocResult<DMAInfo> {
+        let num_pages = layout_pages(&layout);
+        let vaddr_raw =
+            global_allocator().alloc_pages(num_pages, PAGE_SIZE_4K.max(layout.align()))?;
+        let vaddr = VirtAddr::from(vaddr_raw);
+        self.update_flags(
+            vaddr,
+            num_pages,
+            MappingFlags::READ | MappingFlags::WRITE | MappingFlags::UNCACHED,
+        )?;
+        Ok(DMAInfo {
+            cpu_addr: unsafe { NonNull::new_unchecked(vaddr_raw as *mut u8) },
+            bus_addr: virt_to_bus(vaddr),
+        })
+    }
+
+    fn update_flags(
+        &mut self,
+        vaddr: VirtAddr,
+        num_pages: usize,
+        flags: MappingFlags,
+    ) -> AllocResult<()> {
+        let expand_size = num_pages * PAGE_SIZE_4K;
+        axmm::kernel_aspace()
```

Comment:
Let it returns `axmm::kernel_aspace()...map_err()` directly, and remove `?` and `Ok(())`.